### ISSUE TITLE
performance differences between seemingly equivalent models in Brainscript and Python

### ISSUE THREAD

**Issue by pnpetkov** [Microsoft/CNTK#2583]
I am porting an ASR acoustic model Brainscript configuration to Python. It's a 10 layer CNN designed "by the book". The BrainScript version gives very good performance - WER=18 %. An equivalent (imho) model in Python makes it to WER = 22 %. Already in the beginning of the training process I can see from the log file that, for both training and validation data, the Python version is lagging behind.

The optimizers I use are SGD ( BrainScript ) and momentum_sgd (Python), where in both cases the momentum is the same (0.9). Pretty much all of the parameters are the same, except possibly the data randomization settings. At least I can't see any other differences. My specific question is how to "translate" the BrainScript code below to Python.

reader=[
readerType=Kaldi2Reader
readMethod=blockRandomize
frameMode=true
nbruttsineachrecurrentiter=1
miniBatchMode=Partial
randomize=1000000
verbosity=0
features=[...
labels=[...
]

In the Python version I use HTKMLF and HTKfeatures deserializers and the minibatch source is set-up as follows:
....
fd=HTKFeatureDeserializer ...
ld=HTKMLFDeserializer ...
return MinibatchSource([fd,ld], max_samples=mx_samples, frame_mode=frm_mod, randomize=True)

I have played around with the parameters randomization_window_in_chunks and randomization_window_in_samples and observe some variability but it is not "significant". 

More generally, where else should I look at to try and resolve the issue. I am assuming that there is no physical constraint that would prevent me from getting the same or higher performance with the Python 
interface.

Thank you very much!

**Comment by rhy-ama** [comment]
Where functionality /API is available in both, Python will never match Brainscript performance due to additional - and contrary to what is affirmed by the team - significant (accumulated over running times) overhead.

**Comment by cha-zhang** [comment]
There should be no difference in the accuracy of BrainScript and Python. Could you first verify if the data you load are identical, buy turning off all randomization?

**Comment by pnpetkov** [comment]
Thanks for the response! I have continued working on the problem but am not seeing the answer yet. Here is what I have done.

i) I verified that the features (Kaldi format) for BrainScript and (HTK format) for Python are indeed the same. I have not yet run an experiment without randomization in BrainScript though.

ii) Switching randomization off in Python leads to a degradation in performance beyond what I already mentioned in the previous post.

iii) I can see that the BrainScript code takes up about twice as much space on the GPU as the Python version. I wonder if that may be an indication that I am not randomizing over the same amount of data.

iv) Playing with the different optimizers/learners, momentum_SGD makes the best start so I stick with it. I wonder though if it is equivalent to the SGD in BrainScript.

#########################################
Here is something that may account for the differences although I am not seeing it yet. I wrote my own training loop as it gives me the flexibility I was looking for. That in turn required that I use FULL_DATA_SWEEP instead of INFINITELY_REPEAT.

###From the top-level script:

        train_source = create_mb_source(scp_tr, mlf_tr, label_map, cntk.io.FULL_DATA_SWEEP, frm_mode, trn_len, feature_dim, num_classes, context) 
        vld_source = create_mb_source(scp_dt, mlf_dt, label_map, cntk.io.FULL_DATA_SWEEP, frm_mode, 0, feature_dim, num_classes, context) 

###The minibatch source creation:

def create_mb_source(features_file, labels_file, label_mapping_file, mx_samples, frm_mod, trn_len, feature_dim, num_classes, context):
    for file_name in [features_file, labels_file, label_mapping_file]:
        if not os.path.exists(file_name):
            raise RuntimeError("File '%s' does not exist." % (file_name))

        fd = HTKFeatureDeserializer(StreamDefs(
            amazing_features = StreamDef(shape=feature_dim, context=(context,context), scp=features_file)))
   
    ld = HTKMLFDeserializer(label_mapping_file, StreamDefs(
        awesome_labels = StreamDef(shape=num_classes, mlf=labels_file)))

    return MinibatchSource([fd,ld], max_samples=mx_samples, frame_mode=frm_mod, randomize=True)

### From the training:
    ## Train through a loop over minibatch count
    state_mbs_vld=vld_source.get_checkpoint_state()
    state_mbs_trn=train_source.get_checkpoint_state()
    vld_res_epo_old=100                                           # Assuming can't be worse than 100% ;-)
    lr_tmp=learner.learning_rate()
    keep_running=True
    epoch=0                                                       # Keep track of effective number of epochs 
    while keep_running:
    #for epoch in range(max_epochs):

        if ( epoch==0 and os.path.exists( os.path.join(model_path, "best_model_state"))==True ): # Check whether possible to preload a model
            print('\n Loading last best state and continuing from there. \n')
            trainer.restore_from_checkpoint(os.path.join(model_path, "best_model_state"))
            # Possibly do something to come up to speed with the learning rate. 

        # Training 
        train_source.restore_from_checkpoint(state_mbs_trn)
        while True:
            data_tr=train_source.next_minibatch(minibatch_size, input_map) 
            
            if bool(data_tr)==False:    # No samples delivered from MB source
                break
            
            trainer.train_minibatch(data_tr)
            
            if data_tr[network['label']].num_samples<minibatch_size:   # Significant number of samples from MB source smaller than MB size
                break
 
        # Validation
        vld_result=0
        vld_source.restore_from_checkpoint(state_mbs_vld)          # Restore the validation source to beginning
        vld_smp=0
        while True:
            data_vld=vld_source.next_minibatch(minibatch_size, input_map_vld) # Extract minibatch
            if bool(data_vld)==False:    # No samples delivered from MB source
                break
            eval_error=trainer.test_minibatch(data_vld)                   # Average vld MB error
            vld_smp_now=data_vld[network['label']].num_samples
            vld_smp=vld_smp+vld_smp_now
            vld_result=vld_result+eval_error*vld_smp_now
            if vld_smp_now<minibatch_size:
                break

#### The rest of this function writes a summary, checks if there is an improvement on the validation data and if yes, saves the checkpoint and the model and starts the next epoch. If no improvement, then it split the learning rate, reloads the latest saved checkpoint and starts the next epoch.

I'll be grateful to receive any pointers!

**Comment by pnpetkov** [comment]
An update! 

Started using the HTK deserializer from BrainScript on exactly the same data as the Python code works on. Still there is an advantage for BrainScript over Python. Switching randomization off in both versions makes things worse in absolute terms but the relation was preserved. This was for a 10-layer CNN.

At this point I started comparing on DNN architectures and noticed something similar. Moreover the Python code was slower in reaching the same performance levels after each epoch. It was at this point I started thinking that the issue can be in the optimizer/learner settings. 

I tried to make the settings as similar as possible. So, I used "momentum_as_time_constant_schedule(2429.8)" where 2429.8 is the value BrainScript logs based on an effective momentum of 0.9. Until that point I used "momentum_schedule(0.9)". Immediately the DNN started progressing (reducing the cost) much faster. 

The same "trick" didn't quite do it for the CNN architecture although there is a small improvement up-front. How different are BrainScript's SGD and Python's momentum_SGD? 

Also, if I use the equation: momentumAsTimeConstant = -minibatchSize / ln (momentumPerMB) with minibatchSize =256 and momentumPerMB=0.9, I end up with 5594.7  (https://github.com/Microsoft/CNTK/wiki/Tutorial2). This is more two times larger than the value BrainScript logs for the same parameter values. Why is that?

**Comment by pnpetkov** [comment]
An update! 

Started using the HTK deserializer from BrainScript on exactly the same data as the Python code works on. Still there is an advantage for BrainScript over Python. Switching randomization off in both versions makes things worse in absolute terms but the relation was preserved. This was for a 10-layer CNN.

At this point I started comparing on DNN architectures and noticed something similar. Moreover the Python code was slower in reaching the same performance levels after each epoch. It was at this point I started thinking that the issue can be in the optimizer/learner settings. 

I tried to make the settings as similar as possible. So, I used "momentum_as_time_constant_schedule(2429.8)" where 2429.8 is the value BrainScript logs based on an effective momentum of 0.9. Until that point I used "momentum_schedule(0.9)". Immediately the DNN started progressing (reducing the cost) much faster. 

The same "trick" didn't quite do it for the CNN architecture although there is a small improvement up-front. How different are BrainScript's SGD and Python's momentum_SGD? 

Also, if I use the equation: momentumAsTimeConstant = -minibatchSize / ln (momentumPerMB) with minibatchSize =256 and momentumPerMB=0.9, I end up with 5594.7  (https://github.com/Microsoft/CNTK/wiki/Tutorial2). This is more two times larger than the value BrainScript logs for the same parameter values. Why is that?

**Comment by pnpetkov** [comment]
An update! 

Started using the HTK deserializer from BrainScript on exactly the same data as the Python code works on. Still there is an advantage for BrainScript over Python. Switching randomization off in both versions makes things worse in absolute terms but the relation was preserved. This was for a 10-layer CNN.

At this point I started comparing on DNN architectures and noticed something similar. Moreover the Python code was slower in reaching the same performance levels after each epoch. It was at this point I started thinking that the issue can be in the optimizer/learner settings. 

I tried to make the settings as similar as possible. So, I used "momentum_as_time_constant_schedule(2429.8)" where 2429.8 is the value BrainScript logs based on an effective momentum of 0.9. Until that point I used "momentum_schedule(0.9)". Immediately the DNN started progressing (reducing the cost) much faster. 

The same "trick" didn't quite do it for the CNN architecture although there is a small improvement up-front. How different are BrainScript's SGD and Python's momentum_SGD? 

Also, if I use the equation: momentumAsTimeConstant = -minibatchSize / ln (momentumPerMB) with minibatchSize =256 and momentumPerMB=0.9, I end up with 5594.7  (https://github.com/Microsoft/CNTK/wiki/Tutorial2). This is more two times larger than the value BrainScript logs for the same parameter values. Why is that?
